### PR TITLE
[SPARK-24985][SQL][WIP] Fix OOM in Full Outer Join in case of data skew

### PR DESCRIPTION
**Pull request by sujithjay**
## What issue does this pull request address ?
JIRA: [https://issues.apache.org/jira/browse/SPARK-24985](https://issues.apache.org/jira/browse/SPARK-24985)
In the case of Full Outer Joins of large tables, in the presence of data skew around the join keys for either of the joined tables, OOMs exceptions occur. While its possible to increase the heap size to workaround, Spark should be resilient to such issues as skews can happen arbitrarily.

## What changes were proposed in this pull request?

#16909 introduced `ExternalAppendOnlyUnsafeRowArray` & changed `SortMergeJoinExec` to use `ExternalAppendOnlyUnsafeRowArray` for every join, except 'Full Outer Join'. This PR makes changes to make 'Full Outer Joins' to use `ExternalAppendOnlyUnsafeRowArray`.

#### Unit testing
- Changed a test-case in `JoinSuite`.
- Existing tests in `OuterJoinSuite` were used to verify correctness.

#### Stress testing
- This is still a work in progress. I plan to verify this patch using a production workload.

### CODE REVIEW

**Comment by sujithjay** [comment]
Hi @tejasapatil, @viirya, @hvanhovell,  & @kiszk, can you please review this pull request?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 1]
```diff
@@ -1099,7 +1111,7 @@ private class SortMergeFullOuterJoinScanner(
 
   def advanceNext(): Boolean = {
     // If we already buffered some matching rows, use them directly
-    if (leftIndex <= leftMatches.size || rightIndex <= rightMatches.size) {
+    if (leftIndex <= leftMatches.length || rightIndex <= rightMatches.length) {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Why did you change size -> length?

Comment (sujithjay) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: I changed the type of `leftMatches` & `rightMatches` from `ArrayBuffer[InternalRow]` to `ExternalAppendOnlyUnsafeRowArray`. `ExternalAppendOnlyUnsafeRowArray` exposes a `length` method, instead of `size` method.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 2]
```diff
@@ -975,8 +979,10 @@ private class SortMergeFullOuterJoinScanner(
 
   private[this] var leftIndex: Int = 0
   private[this] var rightIndex: Int = 0
-  private[this] val leftMatches: ArrayBuffer[InternalRow] = new ArrayBuffer[InternalRow]
-  private[this] val rightMatches: ArrayBuffer[InternalRow] = new ArrayBuffer[InternalRow]
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: We can remove `scala.collection.mutable.ArrayBuffer` import now. Seems no other places use it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 3]
```diff
@@ -1028,23 +1034,23 @@ private class SortMergeFullOuterJoinScanner(
     rightIndex = 0
 
     while (leftRowKey != null && keyOrdering.compare(leftRowKey, matchingKey) == 0) {
-      leftMatches += leftRow.copy()
+      leftMatches.add(leftRow.copy().asInstanceOf[UnsafeRow])
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: `ExternalAppendOnlyUnsafeRowArray` will handle copy, don't need to do another copy here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 4]
```diff
@@ -1058,31 +1064,37 @@ private class SortMergeFullOuterJoinScanner(
...
-      while (rightIndex < rightMatches.size) {
-        joinedRow(leftMatches(leftIndex), rightMatches(rightIndex))
-        if (boundCondition(joinedRow)) {
-          leftMatched.set(leftIndex)
-          rightMatched.set(rightIndex)
+    val leftMatchesIterator = leftMatches.generateIterator(leftIndex)
+
+    while (leftMatchesIterator.hasNext) {
+      val leftCurRow = leftMatchesIterator.next()
+      val rightMatchesIterator = rightMatches.generateIterator(rightIndex)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Can we keep the scanning left and right iterators? Because if they are spilled, obtaining the iterator from spilled data needs to loop over spill writers and create readers. We may avoid calling `generateIterator` every time for obtaining the iterators. However it might make the code a bit complicated than now.

Comment (sujithjay) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Hi @viirya, 
Thank you for reviewing the code. 

I agree the code will be a bit complicated if we had to keep scanning the iterators. In this particular case, I was following a pattern observed throughout the rest of the class.
For eg., https://github.com/apache/spark/blob/35f7f5ce83984d8afe0b7955942baa04f2bef74f/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala#L307-L329

Having said that, I do feel the penalty for following a similar approach in case of full outer joins could be higher. I will try & see what I can do.

Comment (sujithjay) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Hi @viirya ,
After some deliberation, I figure it would not be possible to avoid the reinitialisation of the right iterator. Please share your thought on this.
